server/src/main/java/org/apache/iotdb/db/query/executor/LastQueryExecutor.java

```diff
@@ -127,103 +127,64 @@ protected TimeValuePair calculateLastPairForOneSeries(
-                timeseriesMetadataStats.getEndTime(),
-                timeseriesMetadataStats.getLastValue(),
-                tsDataType);
-            break;
-          } else {
-            List<ChunkMetadata> chunkMetadataList = timeseriesMetadata.loadChunkMetadataList();
-            if (!chunkMetadataList.isEmpty()) {
-              ChunkMetadata lastChunkMetaData = chunkMetadataList.get(chunkMetadataList.size() - 1);
-              Statistics chunkStatistics = lastChunkMetaData.getStatistics();
-              resultPair =
-                  constructLastPair(
-                      chunkStatistics.getEndTime(), chunkStatistics.getLastValue(), tsDataType);
-              break;
-            }
-          }
-        }
-      }
-    }
+        QueryResourceManager.getInstance().getQueryDataSource(seriesPath, context, filter);
 
-    long version = 0;
-    for (TsFileResource resource : unseqFileResources) {
-      if (resource.getEndTime(seriesPath.getDevice()) < resultPair.getTimestamp()) {
-        continue;
-      }
-      TimeseriesMetadata timeseriesMetadata =
-          FileLoaderUtils
-              .loadTimeSeriesMetadata(resource, seriesPath, context, null, deviceMeasurements);
-      if (timeseriesMetadata != null) {
-        for (ChunkMetadata chunkMetaData : timeseriesMetadata.loadChunkMetadataList()) {
-          if (chunkMetaData.getEndTime() > resultPair.getTimestamp()
-              || (chunkMetaData.getEndTime() == resultPair.getTimestamp()
-              && chunkMetaData.getVersion() > version)) {
-            Statistics chunkStatistics = chunkMetaData.getStatistics();
-            resultPair =
-                constructLastPair(
-                    chunkStatistics.getEndTime(), chunkStatistics.getLastValue(), tsDataType);
-            version = chunkMetaData.getVersion();
-          }
-        }
-      }
-    }
+    LastPointReader lastReader = new LastPointReader(
+        seriesPath, tsDataType, deviceMeasurements, context, dataSource, Long.MAX_VALUE, filter);
+    TimeValuePair resultPair = lastReader.readLastPoint();
 
-    // Update cached last value with low priority
-    if (lastCacheEnabled) {
-      IoTDB.metaManager.updateLastCache(seriesPath,
-              resultPair, false, Long.MIN_VALUE, node);
+    // Update cached last value with low priority unless "FROM" expression exists
+    if (lastCacheEnabled && filter == null) {
+      IoTDB.metaManager.updateLastCache(
+          seriesPath, resultPair, false, Long.MIN_VALUE, node);
     }
     return resultPair;
   }
 
-  private static TimeValuePair constructLastPair(long timestamp, Object value,
-                                                 TSDataType dataType) {
-    return new TimeValuePair(timestamp, TsPrimitiveType.getByType(dataType, value));
+  private static boolean satisfyFilter(Filter filter, TimeValuePair tvPair) {
+    if (filter != null &&
+        filter.satisfy(tvPair.getTimestamp(), tvPair.getValue().getValue())) {
+      return true;
+    }
```

Comment:
if the filter is null, shoudn't we return true?

Reply:
Fixed